### PR TITLE
Adding publishing to Central Repository using a github action

### LINKED ISSUE #68: Switch from Jitpack to Maven Central

### LINKED ISSUE #126: Automated GH Release Tag
We need a GH Action workflow to generate the GH releases.

Similar to https://github.com/TBD54566975/tbdex-js/pull/61

### LINKED ISSUE #128: Setup artifacts publishing
We need to have an easy automated workflow to publish artifacts to the corresponding sdk registry (npm, maven, crates etc.), and also publish to the internal Artifactory registry too

### PR DESCRIPTION
**Pull request by andresuribe87**
# Overview

This PR fixes #68, fixes #126, and fixes #128 by configuring gradle, the repo, and a CI action to publish the libraries from this repo to maven central. 

# Description

This PR includes. 
* Updates to the README file on how to import this library, and how to realease it. The current `groupId` we've settled on is `xyz.block`, but that's likely to change (see the convo at https://github.com/TBD54566975/open-source-programs/issues/95). When the time comes, we'll update it. Alternatively, we can delay this PR until that's ready.
* Adding a `publish.yml` file. This workflow will trigger when a new release is cut. It will release the version specified to maven central. See the README updates for more details. 
* Changes to gradle file such that:
    * All submodules are included in the root module, importable as `web5`.
    * Publications now include details of the POM file.
    * Publications will be signed using a PGP key. (note that I created the key previously by following the instructions in https://central.sonatype.org/publish/requirements/gpg/#credentials).

# How Has This Been Tested?

I tested the CI action by making sure that you can import `xyz.block:web5:test-SNAPSHOT` from a separate repo. Additionally, the jar's can be seen from https://s01.oss.sonatype.org/#welcome, after logging in as `tbd-releases` (the password is in the shared 1password vault).

All the commands in the README were tested as well to behave as expected.

## References
* The main guide: https://central.sonatype.org/publish/publish-gradle/
* And the README of the gradle plugin used: https://github.com/Codearte/gradle-nexus-staging-plugin/#quick-start

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -13,8 +13,38 @@ This repo contains 4 jvm packages:
...
+
+## Maven Central
+
+When pulling from Maven Central, you can pull the entire library or just a single module. Examples of both are shown
+below. Please note that you need to add the repositories shows below to your `build.gradle.kts` file. This is because
+the libraries that we depend on are hosted in places
+
+```kt
+repositories {
+  maven("https://jitpack.io")
```
Comment (jiyoontbd) [inline README.md:27]: i think we typically have `mavenCentral()` inside this block?

File: build.gradle.kts [hunk 2]
```diff
@@ -12,20 +12,24 @@ plugins {
   `maven-publish`
   id("org.jetbrains.dokka") version "1.9.0"
   id("org.jetbrains.kotlinx.kover") version "0.7.3"
+  signing
```
Comment (jiyoontbd) [inline build.gradle.kts:15]: what is this used for?

File: build.gradle.kts [hunk 3]
```diff
@@ -12,20 +12,24 @@ plugins {
...
+  api(project(":common"))
+  api(project(":credentials"))
+  api(project(":crypto"))
+  api(project(":dids"))
 }
 
 allprojects {
-  version = "0.0.9"
-  group = "web5"
+  group = "xyz.block"
```
Comment (jiyoontbd) [inline build.gradle.kts:32]: wao, fancy group name. is this to be the groupId for tbdex-kt as well?

File: build.gradle.kts (hunk 2 above)
Comment (jiyoontbd) [reply build.gradle.kts:15]: nvm found it

File: build.gradle.kts (hunk 3 above)
Comment (andresuribe87) [reply build.gradle.kts:32]: There's a convo in https://github.com/TBD54566975/open-source-programs/issues/95

But I don't know if that includes tbdex-kt. A decision will need to be made.

File: README.md (hunk 1 above)
Comment (andresuribe87) [reply README.md:28]: Added!

**Review by jiyoontbd — APPROVED** [review]
curious if you tried publishing to sonartype staging - [looks like](https://github.com/TBD54566975/web5-kt/actions/runs/7133809818) it was successful? :)

**Comment by andresuribe87** [comment]
> curious if you tried publishing to sonartype staging - [looks like](https://github.com/TBD54566975/web5-kt/actions/runs/7133809818) it was successful? :)

Yes I did! You can see in the PR description more details

File: README.md [hunk 4]
```diff
@@ -13,11 +13,43 @@ This repo contains 4 jvm packages:
 
 # Quickstart
 
-All the web5 libraries are published on [JitPack](https://jitpack.io). To start simply add the following to your
-`gradle.build.kts` file:
+You can add this library you your project using Gradle or Maven. There are two ways to do so. The first is pulling the
```
Comment (phoebe-lew) [inline README.md:16]:
```suggestion
You can add this library to your project using Gradle or Maven. There are two ways to do so. The first is pulling the
```

File: README.md [hunk 5]
```diff
@@ -13,11 +13,43 @@ This repo contains 4 jvm packages:
...
+package from Maven Central. The second is pulling the package from JitPack.
```
Comment (phoebe-lew) [inline README.md:17]: is there a reason we want to continue supporting both maven central and jitpack?

File: README.md [hunk 6]
```diff
@@ -13,11 +13,43 @@ This repo contains 4 jvm packages:
...
+
+## Maven Central
+
+When pulling from Maven Central, you can pull the entire library or just a single module. Examples of both are shown
+below. Please note that you need to add the repositories shows below to your `build.gradle.kts` file. This is because
+the libraries that we depend on are hosted in places
```
Comment (phoebe-lew) [inline README.md:23]: nit: missing end of sentence?

File: README.md [hunk 7]
```diff
@@ -13,11 +13,43 @@ This repo contains 4 jvm packages:
...
+
+```kt
+repositories {
+  mavenCentral()
+  maven("https://jitpack.io")
```
Comment (phoebe-lew) [inline README.md:28]: do we need this if we're pulling from maven central?

File: README.md [hunk 8]
```diff
@@ -13,11 +13,43 @@ This repo contains 4 jvm packages:
...
+```
+
+## JitPack
+
+You can also pull the jars for this library from [JitPack](https://jitpack.io). To start simply add the following to your
+`build.gradle.kts` file:
 
 ```kotlin
 repositories {
+  mavenCentral()
```
Comment (phoebe-lew) [inline README.md:52]: needed if just pulling from jitpack?

File: README.md [hunk 9]
```diff
@@ -52,12 +84,44 @@ in [GitHub Pages](https://tbd54566975.github.io/web5-kt/docs/htmlMultiModule/cre
...
+
 To build and run test just run:
 
 ```bash
 ./gradlew build --console=rich
 ```
 
+## Releasing
+
+In order to release to Central Repository, simply cut a new release tag in the Github UI. There is a configured [Github
```
Comment (phoebe-lew) [inline README.md:105]: Does this allow for bumping individual packages, or will this bump everything?

Comment (phoebe-lew) [reply README.md:105]: Does this action release to both jitpack and maven central?

File: README.md (hunk 5 above)
Comment (andresuribe87) [reply README.md:17]: I didn't want to replace immediately in this PR. Instead, strip it away once we are 100% sure that maven is working as expected.

File: README.md (hunk 7 above)
Comment (andresuribe87) [reply README.md:28]: Yes, some of our transitive deps still come from jitpack.

File: README.md (hunk 8 above)
Comment (andresuribe87) [reply README.md:52]: Yes, due to transitive deps as well :)

File: README.md (hunk 9 above)
Comment (andresuribe87) [reply README.md:105]: > Does this allow for bumping individual packages, or will this bump everything?

It does not allow for bumping individual packages. We'll have to bump everything.

> Does this action release to both jitpack and maven central?

Jitpack works differently than maven. You don't need to "release". Instead, it builds a JAR on the fly when you ask for a specific version. You can refer to which JAR by using a PR, tag, or commit.

File: README.md (hunk 6 above)
Comment (andresuribe87) [reply README.md]: Corrected.

File: .github/workflows/publish.yml [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+on:
+  workflow_dispatch:
+    inputs:
+      version:
+        description: 'Version to publish. For example "0.0.1"'
+        required: true
+        default: '0.0.0'
+  push:
+    tags:
+      - v*
```
Comment (leordev) [inline .github/workflows/publish.yml:11]: Creating a tag will trigger this publishing, I think what you actually want is to trigger when a new GH release is created or, even more strict, published. ([reference](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#release))

Comment (andresuribe87) [reply .github/workflows/publish.yml:11]: Awesome, will give it a go.

File: build.gradle.kts [hunk 11]
```diff
@@ -166,3 +208,57 @@ subprojects {
...
+        }
+        scm {
+          connection.set("scm:git:git@github.com:TBD54566975/web5-kt.git")
+          developerConnection.set("scm:git:ssh:git@github.com:TBD54566975/web5-kt.git")
+          url.set("https://github.com/TBD54566975/web5-kt")
+        }
+      }
+    }
+  }
+}
```
[the remark this answers is not in the document]
Comment (andresuribe87) [reply build.gradle.kts:248]: I tried a couple of different strategies, but gradle is very opionated and unfortunately I couldn't get it to work :(

**Review by phoebe-lew — APPROVED** [review]

**Pull request merged**